File: 009-HarvardX-CS50/pset1/credit/credit.c
#include <stdio.h>
#include <math.h>
#include <cs50.h>


int main(void)
{
    long card;
    int size;
    //Card number input
    card = get_long("Number: ");
    size = (int)floor(log10(card)) + 1;


    //Length validation
    if (size == 13 || size == 15 || size == 16)
    {
        // Converting card number to array of int
        int n[size];
        for (int i = 0; i < size; i++)
        {
            n[i] = card % 10;
            card = round(card / 10);
            //printf("%d\n",n[i]);
        }
        // Luhn algorithm
        int sumOdd = 0;
        int sumEven = 0;
        for (int i = 1; i < size; i += 2)
        {
            sumOdd += (n[i] * 2 % 10) + floor(n[i] * 2 / 10);
        }
        for (int i = 0; i < size; i += 2)
        {
            sumEven += n[i];
        }
        //printf("%d\n",sumOdd);
        //printf("%d\n",sumEven);
        int luhn = sumOdd + sumEven;
        printf("Size: %d\n n[size]:%d\n n[size-1] %d\n luhn %d\n", size, n[size - 1], n[size - 2], luhn);
        if ((luhn % 10) == 0)
        {
            if ((size == 15) && (n[size - 1] == 3) && ((n[size - 2] == 4) || (n[size - 2] == 7)))
            {
                printf("AMEX\n");
            }
            else if ((size == 16) && (n[size - 1] == 5) \
    && (n[size - 2] == 1 || n[size - 2] == 2 || \
    n[size - 2] == 3 || n[size - 2] == 4 || n[size - 2] == 5))
            {
                printf("MASTERCARD\n");
            }
            else if (((size == 13) || (size == 16)) && (n[size - 1] == 4))
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: 009-HarvardX-CS50/pset6/credit/credit.py
from cs50 import get_int
from math import floor

# Getting input from user
card = get_int("Number: ")
size = len(str(card))
card_str = str(card)
print(size)

# Checking length
if size in [13, 15,16]:
    # Convert int to array of digits
    n = []
    for i in range(size):
        n.append(card % 10);
        card = floor(card / 10);
    sumOdd = 0
    sumEven = 0
    # Calculation in accordance with Luhn's algorithm
    for i in range(1, size, 2):
        sumOdd += (n[i] * 2 % 10) + floor(n[i] * 2 / 10)
    for i in range(0, size, 2):
        sumEven += n[i]
    luhn = sumOdd + sumEven
    print(n)
    # if card number is valid (luhn), then check special parameters - length and first digits
    if ((luhn % 10) == 0):
        if ((size == 15) and (n[size - 1] == 3)) and ((n[size - 2] == 4) or (n[size - 2] == 7)):
            print("AMEX")
        elif ((size == 16) and (n[size - 1] == 5)
    and (n[size - 2] == 1 or n[size - 2] == 2 or
    n[size - 2] == 3 or n[size - 2] == 4 or n[size - 2] == 5)):
            print("MASTERCARD")
        elif (((size == 13) or (size == 16)) and (n[size - 1] == 4)):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")
else:
    print("INVALID")


